src/relation_handlers.py

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Pebble handler wrapper classes for this charm"""
+import logging
+from typing import Callable, Optional
+
+import ops.charm
+import ops.model
+from ops.charm import RelationEvent
+from ops.framework import (
+    Handle,
+
+    BoundEvent,
+    EventBase,
+    EventSource,
+    Object,
+    StoredState,
+    ObjectEvents,
+)
+from ops.pebble import Layer
+from ops_sunbeam.relation_handlers import RelationHandler
+from ops_sunbeam.guard import BlockedExceptionError, WaitingExceptionError, guard
+from serialized_data_interface import (
+    SerializedDataInterface,
+    NoCompatibleVersions,
+    NoVersionsListed,
+    get_interface,
+)
+
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    UnknownStatus,
+    WaitingStatus,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SdiErrorEvent(RelationEvent):
+    """Event triggers when an SDI relation catches an exception.
+
+    This is a base object and not meant to be used directly.
+    """
+
+    def __init__(
+            self,
+            handle: ops.framework.Handle,
+            relation: ops.Relation,
+            app: Optional[ops.model.Application] = None,
+            unit: Optional[ops.model.Unit] = None,
+            message: Optional[str] = None,
+    ):
+        super().__init__(handle=handle, relation=relation, app=app, unit=unit)
+        # TODO: Needed?  Or does the event already have enough?
+        self.message = message
+
+
+class SdiNoVersionsListedErrorEvent(SdiErrorEvent):
+    """Event triggers when an SDI relation catches a NoVersionsListed exception."""
+
+
+class SdiNoCompatibleVersionsEvent(SdiErrorEvent):
+    """Event triggers when an SDI relation catches a NoCompatibleVersions exception."""
+
+
+class SdiRelationProviderEvents(ObjectEvents):
+    """Container for events for SDI relation errors."""
+    no_versions_listed = EventSource(SdiNoVersionsListedErrorEvent)
+    no_compatible_versions = EventSource(SdiNoCompatibleVersionsEvent)
+
+
+class KubeflowProfilesProvides(Object):
```

Comment:
`KubeflowProfilesProvides` essentially turns the SDI-style relation into a typical Charm library.  This isn't something we'd need to add for every relation, just something because we haven't modernized this relation into a library yet.